### ISSUE TITLE
Print specific pages

### ISSUE THREAD

**Issue by maxgolopolosov1990** [stephanrauh/ngx-extended-pdf-viewer#243]
Hi, how can I print only some pages? Ie I need the user to be able to print only certain pages

**Comment by stephanrauh** [comment]
ngx-extended-pdf-viewer doesn't really print itself. It just prepares a couple of images (one image per page) and sends them to the browser's print dialog.

So no matter what you do, there's always the print dialog of the OS. You can chose the page range in this dialog. Do you see any advantages in selecting the page range earlier?

![image](https://user-images.githubusercontent.com/3045767/77857036-9b08da00-71fb-11ea-9941-3065a9371c4f.png)

**Comment by maxgolopolosov1990** [comment]
Thanks for the answer. But that doesn't suit me. For example, I need to prevent the user from printing the last page in my PDF, how can I achieve this result?

**Comment by stephanrauh** [comment]
If your user really wants to print the last page, there's no way to stop them from doing the. The page is in memory, so there's always a way to print it. They can use screenshot if nothing else helps.

**Comment by stephanrauh** [comment]
But let's assume you've got a friendly user. I think it's possible to have pdf.js render a page range instead of rendering every page. Basically, we'd have to path this function: https://github.com/mozilla/pdf.js/blob/c12ea21c14478df9a4f4bc10a078d902c174a6d6/web/pdf_print_service.js#L150

**Comment by maxgolopolosov1990** [comment]
I agree with you, then a slightly different question, for example, when you click the Print button - a request was sent to my server and the file returned, and after that it immediately got out the print window?

**Comment by stephanrauh** [comment]
Before adding this feature to ngx-extended-pdf-viewer, let me ponder about it. On the plus side, this feature makes a lot of sense with huge documents. Rendering the images can take a lot of time.

In the meantime, you can implement this feature yourself by monkey-patching the method.

**Comment by maxgolopolosov1990** [comment]
> I agree with you, then a slightly different question, for example, when you click the Print button - a request was sent to my server and the file returned, and after that it immediately got out the print window?

how can I do it?

**Comment by stephanrauh** [comment]
I've found this recipe on [StackOverflow](https://stackoverflow.com/questions/16239513/print-pdf-directly-from-javascript). I didn't try it yet, but it looks like a reasonable solution. You might also want to add an `onafterprint` callback to get rid of the iFrame again.

```javascript
printPdf = function (url) {
  var iframe = this._printIframe;
  if (!this._printIframe) {
    iframe = this._printIframe = document.createElement('iframe');
    document.body.appendChild(iframe);

    iframe.style.display = 'none';
    iframe.onload = function() {
      setTimeout(function() {
        iframe.focus();
        iframe.contentWindow.print();
      }, 1);
    };
  }

  iframe.src = url;
}
```

**Comment by stephanrauh** [comment]
A programmatic print API has landed with version 3.5.0. It still doesn't prevent the user from hitting CTRL+P or the print button of the toolbar, but it's a start. You can now define which pages you want to print.

The documentation is available at https://pdfviewer.net/print-range.

**Comment by stephanrauh** [comment]
I've added the API to set the print range for the print button and the key binding CTRL+P with version 3.5.1.
